pkg/virt-handler/node-labeller/node_labeller.go

```diff
@@ -0,0 +1,185 @@
+
+import (
+	goerror "errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+	device_manager "kubevirt.io/kubevirt/pkg/virt-handler/device-manager"
+)
+
+const (
+	labelNamespace    = "feature.node.kubernetes.io"
+	labellerNamespace = "node-labeller"
+)
+
+var nodeLabellerVolumePath = "/var/lib/kubevirt-node-labeller"
+
+//NodeLabeller struct holds informations needed to run node-labeller
+type NodeLabeller struct {
+	kvmController *device_manager.DeviceController
+	clientset     kubecli.KubevirtClient
+	host          string
+}
+
+func NewNodeLabeller(kvmController *device_manager.DeviceController, clientset kubecli.KubevirtClient, host string) *NodeLabeller {
+	return &NodeLabeller{
+		kvmController: kvmController,
+		clientset:     clientset,
+
+		host: host,
+	}
+}
+
+//Run runs node-labeller
+func (n *NodeLabeller) Run(stop chan struct{}) error {
+	logger := log.DefaultLogger()
+	logger.Infof("node-labeller is running")
+	kvmPresent := false
+	retries := 0
+	maxRetries := 30
+
+	// check if kvm device is available
+	for {
+		if n.kvmController.NodeHasDevice(device_manager.KVMPath) {
+			kvmPresent = true
+			break
+		}
+
+		select {
+		case <-stop:
+			return goerror.New(fmt.Sprintf("Node-labeller has stopped"))
+		default:
+			retries++
+			if retries > maxRetries {
+				break
+			}
+			time.Sleep(time.Minute)
+		}
+
+	}
+
+	if !kvmPresent {
+		return goerror.New(fmt.Sprintf("Node-labeller cannot work without KVM device."))
+	}
+
+	for {
+		cpuFeatures := make(map[string]bool)
+		cpuModels := make([]string, 0)
+		//parse all informations
+		cpuModels, cpuFeatures, err := getCPUInfo()
+
+		//wait until data from virt-launcher are copied to virt-handler
+		if (len(cpuFeatures) == 0 && len(cpuModels) == 0) || err != nil {
+			time.Sleep(time.Second * 30)
+			continue
+		}
+
+		//get node on which virt-handler is running
+		node, err := n.clientset.CoreV1().Nodes().Get(n.host, metav1.GetOptions{})
+		if err != nil {
+			logger.Infof("node-labeller cannot get node %s", n.host)
+		}
+
+		//prepare new labels
+		newLabels := n.prepareLabels(cpuModels, cpuFeatures)
+		//remove old labeller labels
+		n.removeCPULabels(node, n.getNodeLabellerLabels(node))
+		//add new labels
+		n.addNodeLabels(node, newLabels)
+		//patch node with new labels
+		data := []byte(fmt.Sprintf(`[{"op": "replace", "path": "/metadata/labels", "value": {%s}}, {"op": "replace", "path": "/metadata/annotations", "value": {%s}}]`, convertMapToText(node.Labels), convertMapToText(node.Annotations)))
+		_, err = n.clientset.CoreV1().Nodes().Patch(node.Name, types.JSONPatchType, data)
+		if err != nil {
+			logger.Infof("error during node %s update. %s\n", node.Name, err)
+		}
+		select {
+		case <-stop:
+			return goerror.New(fmt.Sprintf("Node-labeller has stopped"))
+		default:
+			time.Sleep(1 * time.Minute)
```

Comment:
right now this logic is in an infinite loop that executes every minute. So every minute we're pulling in the node from the api server and executing a Patch command on every virt-handler in the cluster.

So, If we have 500 nodes, then that's 1000 requests a minute this logic is generating. Or and average of 16 requests a second.

I see two options here.

1. we can set the labels on the node once and exit. THis would mean if someone mutated the labels they wouldn't get reset.

2. We can create a reconcile loop that uses a node informer watching the specific node associated with this host. The loop would then only get triggered when the specific node in question is mutated.